### ISSUE TITLE
Use the OSGI Extender Pattern for finding migrations

### ISSUE THREAD

**Issue by chrsoo** [flyway/flyway#1102]
In an OSGI environment, the [Extender Pattern with a BundleTracker](http://toedter.com/2010/07/05/osgi-4-2-extender-pattern-and-bundletracker/) is probably a better alternative for finding migrations than classpath scanning. Using the Extender Pattern would resolve issues #1098 and #1007.
# Usecase
1. The `flyway-core` bundle is installed and started in the OSGI container
2. OSGI Bundles that need flyway support adds a Flyway configuration file in
   
   ```
   META-INF/flyway/<name>.properties
   ```
3. When the bundle deployed in OSGI and enters the Starting state, `flyway-core` will scan the bundle for all property file resources in `META-INF/flyway`
4. For each property file found it will lookup the corresponding Flyway `name` Managed Service Configuration
   1. ~~If the a Managed Service Configuration cannot be found a new Managed Service Configuration is created using the content of the found property file resource as default~~
   2. If a Managed Service Flyway Configuration is found it is merged with the property file resource found in the bundle; the Managed Service Flyway Configuration overrides properties found in the property file.
5. From the merged (or default) Flyway configuration a JDBC Driver instance is created using the OSGI Compendium JDBC Service if the Compendium JDC Service is available.
   1. If no driver can be found an attempt to load the driver using the Bundle's class loader is made
6. The driver is configured with all configuration properties fromt the merged Flyway configuration that are not prefixed with `flyway.`
7. The driver is used to create an instance of Flyway's default `DriverDataSource`
8. A `Flyway` instance is created and configured with the merged configuration where the driver specific properties have been removed (driver, url, user and password)
9. Migration is done using the Flyway instance
# Managed Service Flyway Configurations

As all resource files in the bundles represent static and default Flyway configurations the an OSGI Managed Service Factory is used to provide environment specific runtime configurations for Flyway. Typically such configurations would contain at least the JDBC URL, username and password.

Managed Service configurations use the factory pid `org.flywaydb.datasource` and individual configurations can thus be defined in the configuration files on the form

```
org.flywaydb.core-<name>.cfg
```
# Example

Given a bundle `security-authz` handling Authorizations in an AuthzDB flyway the following resources are found in the bundle:

```
META-INF
  flyway
    db/migration
       V1__Initial_structure.sql
       V1_1__Populate_table.sql
    authz.properites
```

The file resource `authz.properties` contains

``` properties
flyway.driver = com.mysql.jdbc.Driver
flyway.user = authz_user
flyway.password = CHANGEME
flyway.url = jdbc:mysql://localhost:3306/AuthzDB
flyway.locations = META-INF/flyway/db/migration
```

The Flyway Managed Service Configuration for the AuthzDB is found in 

```
${etc}/com.flywaydb.core-authzdb.cfg
```

... and it overrides the JDBC url and password:

```
flyway.password = secret
flyway.url = jdbc:mysql://authzdb.internal.acme.com:3306/AuthzDB
```

**Comment by chrsoo** [comment]
I have made a first cut at implementing the above but I have problems compiling Flyway. Running maven with
````
mvn install -P-InstallableDBTest -P-CommercialDBTest -P-CommandlinePlatformAssemblies
````
... the build of the `flyway.gradle.plugin`fails due to Groovy's dynamic classloading mechanism; it chooses the new constructor I added to the DriverDataSource class I need to add the JDBC driver instance I load from the OSGI Compendium JDBC service.

It seems to me that the right way to resolve this is to force static typing of the `AbstractFlywayTask.grove` which then (I think) would force the use of the right constructor. This can be fixed using the [`CompileStatic` annotation](http://docs.groovy-lang.org/latest/html/documentation/index.html#compilestatic-annotation) but that in turn requires upgrading to using [Groove 2.0](http://www.h-online.com/open/news/item/Groovy-2-0-introduces-static-type-checking-and-compiling-1628119.html), doing that on the other hand provokes another Groove compilation fault and somewhere here I am starting to think that maybe I should ask somebody who actually knows Gradle, Groove etc for help. Anybody has any input?

Code is available in my [OSGI branch and Gitub fork of Flyway](https://github.com/chrsoo/flyway/tree/osgi-extender)
